sdk/core/azure-core/inc/http/curl/curl.hpp

```diff
@@ -12,79 +12,504 @@
+   */
+  enum class ResponseParserState
+  {
+    StatusLine,
+    Headers,
+    EndOfHeaders,
+  };
+
+  /**
+   * @brief stateful component used to read and parse a buffer to construct a valid HTTP Response.
+   *
+   * It uses an internal string as buffers to accumulate a response token (version, code, header,
+   * etc) until the next delimiter is found. Then it uses this string to keep building the HTTP
+   * Response.
+   *
+   * @remark Only status line and headers are parsed and built. Body is ignored by this component.
+   * A libcurl session will use this component to build and return the HTTP Response with a body
+   * stream to the pipeline.
+   */
+  class ResponseBufferParser {
   private:
-    CurlTransport(const CurlTransport&) = delete;
-    CurlTransport& operator=(const CurlTransport&) = delete;
+    /**
+     * @brief Controls what the parser is expecting during the reading process
+     *
+     */
+    ResponseParserState state;
+    /**
+     * @brief Unique prt to a response. Parser will create an Initial-valid HTTP Response and then
+     * it will append headers to it. This response is moved to a different owner once parsing is
+     * completed.
+     *
+     */
+    std::unique_ptr<Response> m_response;
+    /**
+     * @brief Indicates if parser has found the end of the headers and there is nothing left for the
+     * HTTP Response.
+     *
+     */
+    bool parseCompleted;
 
-    // for every client instance, create a default response
-    std::unique_ptr<Azure::Core::Http::Response> m_response;
-    bool m_firstHeader;
+    /**
+     * @brief This buffer is used when the parsed buffer doesn't contain a completed token. The
+     * content from the buffer will be appended to this buffer. Once that a delimiter is found, the
+     * token for the HTTP Response is taken from this internal sting if it contains data.
+     *
+     * @remark This buffer allows a libcurl session to use any size of buffer to read from a socket
+     * while constructing an initial valid HTTP Response. No matter if the response from wire
+     * contains hundreds of headers, we can use only one fixed size buffer to parse it all.
+     *
+     */
+    std::string internalBuffer;
 
-    CURL* m_pCurl;
+    /**
+     * @brief This method is invoked by the Parsing process if the internal state is set to status
+     * code. Function will get the status-line expected tokens until finding the end of status line
+     * delimiter.
+     *
+     * @remark When the end of status line delimiter is found, this method will create the HTTP
+     * Response. The HTTP Response is constructed by default with body type as Stream.
+     *
+     * @param buffer Points to a memory address with all or some part of a HTTP status line.
+     * @param bufferSize Indicates the size of the buffer.
+     * @return Returns the index of the last parsed position from buffer.
+     */
+    size_t BuildStatusCode(uint8_t const* const buffer, size_t const bufferSize);
```

Comment:
Maybe `const` is a little bit abused here?

But I think you should at least be consistent, if you use `size_t const` here, you should use it for all parameters like `int`, `uint64_t` etc.

Reply:
I agree - I suggest to use const everywhere, except for the value part (and NOT the pointer part) of a function parameter:
```
std::string const& GetName(bool uppercase) const
{
    auto const isNull = m_name == nullptr;
    ...
    auto const size = m_name.size();
    for (auto i = 0; i < size; ++i) ...
}
```
^^^ all good examples and `const` instances should be used.

Bad example: `void f(bool const b)` - const should not be used.
`void f(char const* p)` - good
`void f(char const* const p)` - second const should not be used, should be : `void f(char const* p)`
`void f(char* p)` - good as long as you are modifying `*p` (not the `p` as a pointer, but the contents it's pointing at). In case we don't modify, we should use `void f(char const* p)`.

Reply:
So, in this case, `size_t BuildStatusCode(uint8_t const* buffer, size_t buffer_size)`